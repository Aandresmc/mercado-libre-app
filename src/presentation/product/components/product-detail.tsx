import React from 'react'
import { IProductDetail } from '@/domain/entities/iproduct-detail'
import { ProductButton } from './product-button'
import styles from './product.module.scss'

interface Props {
	product: IProductDetail.Data
}

export const ProductDetail = ({ product }: Props) => {
	return (
		<div className={styles.container_detail}>
			<p>
				{product.condition == 'new' ? 'Nuevo' : 'Buen estado'} - {product.sold_quantity} vendidos
			</p>
			<h1 className={styles.title_product}>{product.title}</h1>
			<h1 className={styles.price_product}>$ {product.price}</h1>
			<ProductButton />
		</div>
	)
}
