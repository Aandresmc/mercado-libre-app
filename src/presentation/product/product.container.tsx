import { useState, useEffect } from 'react'
import { ProductsLayout } from '@/shared/components/layouts/products.layout'
import { ProductView } from './product.view'
import { IProductDetail } from '@/domain/entities/iproduct-detail'
import { getProductByIdUseCase } from '@/domain/usecases/get-product-by-id.usecase'

export interface Props {
	id: string
}
export const Product = ({ id }: Props) => {
	const [product, setProduct] = useState<IProductDetail.Data>()

	useEffect(() => {
		getProductByIdUseCase.call({ id }).then((result) => {
			setProduct(result)
		})
	}, [id])

	return <ProductsLayout children={product && <ProductView product={product} />} />
}
