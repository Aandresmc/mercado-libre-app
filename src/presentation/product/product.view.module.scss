.container_product {
	display: flex;
	padding-block: 2rem;
	padding-inline: 10rem;
	justify-content: flex-start;
	grid-gap: 5rem;
}

.image_product {
	width: 100%;
	height: auto;
}

@media screen and (max-width: 768px) {
	.container_product {
		display: grid;
		padding-inline: 0rem;
		justify-content: center;
	}
	.image_product {
		padding-inline: 0rem;
		padding-block: 0rem;
	}
}
