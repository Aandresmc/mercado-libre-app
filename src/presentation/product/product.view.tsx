import React from 'react'
import Image from 'next/image'
import { IProductDetail } from '@/domain/entities/iproduct-detail'
import styles from './product.view.module.scss'
import { ProductDetail } from './components/product-detail'
import { ProductDescription } from './components/product-description'

export interface Props {
	product: IProductDetail.Data
}
export const ProductView = ({ product }: Props) => (
	<>
		<section className={styles.container_product}>
			<Image
				className={styles.image_product}
				title={'image product'}
				src={product.pictures[0].url}
				alt={`image product ${product.title}`}
				width={'680'} height={'680'}
				priority
			/>
			<ProductDetail product={product} />
		</section>
		<section>
			<ProductDescription product={product} />
		</section>
	</>
)
