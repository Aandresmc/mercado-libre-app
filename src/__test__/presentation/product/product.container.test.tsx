import React from 'react'

import { act, cleanup, render } from '@testing-library/react'
import { Product, Props } from '@/presentation/product/product.container'
import { productDetailMock } from '../../mocks/product-detail.mock'
import mockNextRouter from '../../utils/mock-router'
import { GetProductByIdUseCase } from '@/domain/usecases/get-product-by-id.usecase'

const props: Props = {
	id: '1',
}

const mockGetProductByIdUseCase = jest.spyOn(GetProductByIdUseCase.prototype, 'call')
mockGetProductByIdUseCase.mockImplementation(() => Promise.resolve(productDetailMock))

describe('ProductContainer Component', function () {
	beforeEach(() => {
		mockNextRouter({
			pathname: '/items',
			query: {
				id: props.id,
			},
		})

		jest.clearAllMocks()
	})

	it('should mounted component', () => {
		const { container } = render(<Product id={props.id} />)
		expect(container).toMatchSnapshot()
	})

	it('should have name and price product', async function () {
		const { getByText, asFragment } = await act(async () => render(<Product id={props.id} />))

		expect(asFragment()).toMatchSnapshot()
		expect(getByText(productDetailMock.title)).toBeInTheDocument()
		expect(getByText(`$ ${productDetailMock.price}`)).toBeInTheDocument()

		cleanup()
	})
})
