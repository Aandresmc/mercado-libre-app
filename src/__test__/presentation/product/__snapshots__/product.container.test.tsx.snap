// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductContainer Component should have name and price product 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="navbar"
    >
      <a
        href="/"
      >
        <img
          alt="brand image"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Fbrand.png&w=128&q=75"
          srcset="/_next/image?url=%2Fbrand.png&w=64&q=75 1x, /_next/image?url=%2Fbrand.png&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
      </a>
      <form
        class="container"
      >
        <div
          class="input-group"
        >
          <input
            alt="search product"
            class="form-control"
            name="search"
            placeholder="Nunca dejes de buscar"
            required=""
            type="text"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-secondary"
              type="submit"
            >
              <i
                class="bi bi-search"
              />
            </button>
          </div>
        </div>
      </form>
    </nav>
  </header>
  <main>
    <section
      class="container_product"
    >
      <img
        alt="image product product title"
        class="image_product"
        data-nimg="1"
        decoding="async"
        height="680"
        src="/_next/image?url=http%3Afake-url-img&w=1920&q=75"
        srcset="/_next/image?url=http%3Afake-url-img&w=750&q=75 1x, /_next/image?url=http%3Afake-url-img&w=1920&q=75 2x"
        style="color: transparent;"
        title="image product"
        width="680"
      />
      <div
        class="container_detail"
      >
        <p>
          Nuevo - 1 vendidos
        </p>
        <h1
          class="title_product"
        >
          product title
        </h1>
        <h1
          class="price_product"
        >
          $ 1000
        </h1>
        <button
          class="btn"
          data-testid="btn-buy"
          type="button"
        >
          <span>
            Comprar
          </span>
        </button>
      </div>
    </section>
    <section>
      <div
        class="container_description"
      >
        <h1
          class="title_product"
        >
          Descripcion
        </h1>
        <p>
          description product
        </p>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`ProductContainer Component should mounted component 1`] = `
<div>
  <header>
    <nav
      class="navbar"
    >
      <a
        href="/"
      >
        <img
          alt="brand image"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Fbrand.png&w=128&q=75"
          srcset="/_next/image?url=%2Fbrand.png&w=64&q=75 1x, /_next/image?url=%2Fbrand.png&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
      </a>
      <form
        class="container"
      >
        <div
          class="input-group"
        >
          <input
            alt="search product"
            class="form-control"
            name="search"
            placeholder="Nunca dejes de buscar"
            required=""
            type="text"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-secondary"
              type="submit"
            >
              <i
                class="bi bi-search"
              />
            </button>
          </div>
        </div>
      </form>
    </nav>
  </header>
  <main />
</div>
`;
